### PR TITLE
OCC BRep format

### PR DESCRIPTION
**Pull request by snoyer**
> [Open Cascade BRep format](https://dev.opencascade.org/doc/overview/html/specification__brep_format.html) could be useful for CAD users working with OCC-based procedural generation (eg. [`cadquery`](https://github.com/CadQuery/cadquery), [`build123d`](https://github.com/gumyr/build123d)) to F3D pipelines.
> 
> example file here: https://github.com/tpaviot/pythonocc-demos/blob/master/assets/models/Motor-c.brep

I'm not sure I'm doing it right but I copy-pasted my way through to loading a `.brep` file.
Motor-c brep

The tests don't pass on my machine (including the existing `.stp` file, even though the test files load and are visually correct in the app) but they look like they should so here we go anyway :)

### CODE REVIEW

File: plugins/occt/module/Testing/TestF3DOCCTReader.cxx
Comment (mwestphal) [inline plugins/occt/module/Testing/TestF3DOCCTReader.cxx:29]: testing all format here are not absolutely needed, as you need to add test in `application/testing/CMakeLists.txt` anyway.

File: plugins/occt/module/Testing/TestF3DOCCTReader.cxx
Comment (mwestphal) [inline plugins/occt/module/Testing/TestF3DOCCTReader.cxx:23]: you would need a `reader->SetFileFormat(vtkF3DOCCTReader::FILE_FORMAT::BREP);`  and same for iges.

**Review by mwestphal — CHANGES_REQUESTED** [review]

File: plugins/occt/module/vtkF3DOCCTReader.cxx
Comment (mwestphal) [inline plugins/occt/module/vtkF3DOCCTReader.cxx:276]: not a big fan of a logic based on colors size. What about checking if color tools is available as soon as possible and not allocating colors at all if not needed ?

File: plugins/occt/module/vtkF3DOCCTReader.cxx
Comment (mwestphal) [inline plugins/occt/module/vtkF3DOCCTReader.cxx:602]: a switch case on top may be a better approach, especially if we add more of them in the future.

File: plugins/occt/module/vtkF3DOCCTReader.cxx
Comment (mwestphal) [inline plugins/occt/module/vtkF3DOCCTReader.cxx:505]: you should be able to test (at least cover it) this easily with a test at VTK module level `TestF3DOCCTReader.cxx`

**Review by mwestphal — CHANGES_REQUESTED** [review]
Nice job! needs some changes still

File: plugins/occt/module/vtkF3DOCCTReader.h
Comment (mwestphal) [inline plugins/occt/module/vtkF3DOCCTReader.h:41]: documentation needs an update too, looks for IGES and STEP in the doc.

File: plugins/occt/module/vtkF3DOCCTReader.h
Comment (mwestphal) [inline plugins/occt/module/vtkF3DOCCTReader.h:42]: also add a changelog entry

File: plugins/occt/module/vtkF3DOCCTReader.cxx
Comment (snoyer) [reply plugins/occt/module/vtkF3DOCCTReader.cxx:276]: the only initialization I can see is: https://github.com/f3d-app/f3d/blob/895480c57ae56ed8800f64591317db364aa7e29b/plugins/occt/module/vtkF3DOCCTReader.cxx#L90-L92

then insertions in 2 place, both conditional on `if (this->ColorTool)`:
https://github.com/f3d-app/f3d/blob/895480c57ae56ed8800f64591317db364aa7e29b/plugins/occt/module/vtkF3DOCCTReader.cxx#L151
https://github.com/f3d-app/f3d/blob/895480c57ae56ed8800f64591317db364aa7e29b/plugins/occt/module/vtkF3DOCCTReader.cxx#L257

Is there some costly allocation I'm missing somewhere? 

Or are we talking about refactoring the whole thing to de-tangle the colors from the other stuff happening in between so it can be guarded by a single `if (this->ColorTool)`? because that sound like a lot of work for a janky file format that nobody uses ;)

File: plugins/occt/module/vtkF3DOCCTReader.cxx
Comment (snoyer) [reply plugins/occt/module/vtkF3DOCCTReader.cxx:505]: When I try opening the `.brep` file as `STEP` format to trigger the code block the test picks up the `... vtkF3DOCCTReader.cxx:504    ERR| ...` message as a failure.
Do I need a separate `.cxx` file with some CMake magic to expect failure as success?

File: plugins/occt/module/vtkF3DOCCTReader.cxx
Comment (mwestphal) [reply plugins/occt/module/vtkF3DOCCTReader.cxx:276]: > Is there some costly allocation I'm missing somewhere?

Nope

> Or are we talking about refactoring the whole thing to de-tangle the colors 

Yes. It is just a suggestion, Perfectly ok not to do it :)

File: plugins/occt/module/vtkF3DOCCTReader.cxx
Comment (mwestphal) [reply plugins/occt/module/vtkF3DOCCTReader.cxx:505]: Right, I forgot about that.

> Do I need a separate .cxx file with some CMake magic to expect failure as success?

Hum, we do have some examples of that in the application testing. If you had a invalid.brep (just an empty file) and try opening it with F3D, do you hit this code ?

If that is so, you can copy the `TestInvalidFullScene` test, where the test result does not matter as long as the text is present.

Comment (snoyer) [reply plugins/occt/module/vtkF3DOCCTReader.cxx:505]: empty file seems to do the trick, thanks!

File: plugins/occt/module/Testing/TestF3DOCCTReader.cxx
Comment (mwestphal) [inline plugins/occt/module/Testing/TestF3DOCCTReader.cxx:21]: that wont test in release mode, you need a return.

File: plugins/occt/module/vtkF3DOCCTReader.cxx
Comment (mwestphal) [inline plugins/occt/module/vtkF3DOCCTReader.cxx:273]: please add a comment stating that this may be empty if OCCT does not provide a ColorTool

**Review by mwestphal — CHANGES_REQUESTED** [review]
small changes needed

**Review by mwestphal — APPROVED** [review]
Nice, lets wait for @Meakk review though before merging.

**Review by Meakk — APPROVED** [review]

**Pull request merged**
